contracts/exchangeIssuance/FlashMintWrapped.sol

```diff
@@ -17,220 +17,905 @@
  */
-contract FlashMintWrapped is ReentrancyGuard {
-
-    using Address for address payable;
-    using SafeMath for uint256;
-    using PreciseUnitMath for uint256;
-    using SafeERC20 for IERC20;
-    using SafeERC20 for ISetToken;
-
-    struct WrappedCalldata {
-        address target;
-        uint256 quantity;
-        bytes callData;
-    }
+contract FlashMintWrapped is ReentrancyGuard, Withdrawable {
+  using DEXAdapter for DEXAdapter.Addresses;
+  using Address for address payable;
+  using Address for address;
+  using SafeMath for uint256;
+  using SafeERC20 for IERC20;
+  using SafeERC20 for ISetToken;
 
-    /* ============ Enums ============ */
+  /* ============ Structs ============ */
 
-    /* ============ Constants ============= */
+  struct ComponentSwapData {
+    // unwrapped token version, e.g. DAI
+    address underlyingERC20;
+    // swap data for DEX operation: fees, path, etc. see DEXAdapter.SwapData
+    DEXAdapter.SwapData dexData;
+    // ONLY relevant for issue, not used for redeem:
+    // amount that has to be bought of the unwrapped token version (to cover required wrapped component amounts for issuance)
+    // this amount has to be computed beforehand through the exchange rate of wrapped Component <> unwrappedComponent
+    // i.e. getRequiredComponentIssuanceUnits() on the IssuanceModule and then convert units through exchange rate to unwrapped component units
+    // e.g. 300 cDAI needed for issuance of 1 Set token. exchange rate 1cDAI = 0.05 DAI. -> buyUnderlyingAmount = 0.05 DAI * 300 = 15 DAI
+    uint256 buyUnderlyingAmount;
+  }
 
-    uint256 constant private MAX_UINT96 = 2**96 - 1;
-    address constant public ETH_ADDRESS = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+  struct ComponentWrapData {
+    string integrationName; // wrap adapter integration name as listed in the IntegrationRegistry for the wrapModule
+    bytes wrapData; // optional wrapData passed to the wrapAdapter
+  }
 
-    /* ============ State Variables ============ */
+  /* ============ Constants ============= */
 
-    IController public immutable controller;
+  uint256 private constant MAX_UINT256 = type(uint256).max;
 
-    /* ============ Events ============ */
+  /* ============ Immutables ============ */
 
-    event ExchangeIssue(
-        address indexed _recipient,     // The recipient address of the issued SetTokens
-        ISetToken indexed _setToken,    // The issued SetToken
-        IERC20 indexed _inputToken,     // The address of the input asset(ERC20/ETH) used to issue the SetTokens
-        uint256 _amountInputToken,      // The amount of input tokens used for issuance
-        uint256 _amountSetIssued        // The amount of SetTokens received by the recipient
-    );
+  IController public immutable setController;
+  IDebtIssuanceModule public immutable debtIssuanceModule; // interface is compatible with DebtIssuanceModuleV2
```

Comment:
I don't think this has to be backwards compatible for any previous set Tokens? I assumed we don't have to support the BasicIssuanceModule or WrapAdapters of V1 here

Reply:
I used the debt issuance module here as it's upgradeable versus the basic issuance module. @ckoopmann for the zero ex one, that's iirc the reason why we use the debt issuance module there

Reply:
yes that was also my thinking. it also has hooks and stuff :D basicIssuanceModule is a deadlock in current form, can't be removed

Reply:
Edward mentioned that there is a BIMv2 basically implemented https://github.com/SetProtocol/set-protocol-v2/pull/178 

We might want to go with that one for CMI, to save gas cost etc. 

I'd also change the method names in BIMv2 to be in line with DIMv2

Reply:
sgtm

Reply:
I am resolving this conversation, the decision regarding BIMv2 is not to be made in this PR. 
I renamed debtIssuanceModule to issuanceModule. BIMv2 would be compatible with IDebtIssuanceModule so all options are open